### ISSUE TITLE
[20.03] gpgconf – gpg-agent.conf not respected

### ISSUE THREAD

**Issue by matthew-piziak** [NixOS/nixpkgs#81784]
In 20.03 beta, it is possible to configure gpg-agent as follows:
```
programs.gnupg.agent = {
  enable = true;
  enableSSHSupport = true;
}
```
 
However some configuration options, e.g. `default-cache-ttl`, are not available. Normally these would be placed in `~/.gnupg/gpg-agent.conf`, but as far as I can tell these are no longer respected as they were in NixOS 19.09.

**To Reproduce**
1. Add `programs.gnupg.agent.enable = true` to `configuration.nix` on NixOS 20.03 beta.
2. Place additional options in `~/.gnupg/gpg-agent.conf`.
3. Reboot to ensure everything has been refreshed.
4. Invoke `gpgconf` to see configuration options.

At this point something is clearly wrong, because the `pinentry` entry emitted by `gpgconf` is `/nix/store/2qi2pgxjvlkkxchrp3hsdkxn77szd60a-gnupg-2.2.19/bin/pinentry`, which does not exist in my filesystem.

5. Invoke `gpgconf --list-options gpg-agent`

Here, I'd expect the options to respect my `gpg-agent.conf`, but instead options like `default-cache-ttl` are set to their defaults.

What is the right way to set `default-cache-ttl` on NixOS 20.03 beta?

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.21, NixOS, 20.03beta-436.43ab8e86ca6 (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.2`
 - channels(root): `"nixpkgs-20.03beta-436.43ab8e86ca6"`
 - channels(matt): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`

**Comment by worldofpeace** [comment]
@flokli

**Comment by flokli** [comment]
@matthew-piziak I made gpg-agent a bit more spammy:
```
diff --git a/nixos/modules/programs/gnupg.nix b/nixos/modules/programs/gnupg.nix
index 7a3cb588ee7..65a92aef5ee 100644
--- a/nixos/modules/programs/gnupg.nix
+++ b/nixos/modules/programs/gnupg.nix
@@ -96,7 +96,7 @@ in
     # This overrides the systemd user unit shipped with the gnupg package
     systemd.user.services.gpg-agent = mkIf (cfg.agent.pinentryFlavor != null) {
       serviceConfig.ExecStart = [ "" ''
-        ${cfg.package}/bin/gpg-agent --supervised \
+        ${cfg.package}/bin/gpg-agent -vv --debug --supervised \
           --pinentry-program ${pkgs.pinentry.${cfg.agent.pinentryFlavor}}/bin/pinentry
       '' ];
     };
```

After switching my system to that config, I did a 
```
systemctl --user daemon-reload
systemctl --user restart gpg-agent
```

And watching the log (`journalctl --user -u gpg-agent -n -f`) shows `gpg-agent` is reading from `gpg-agent.conf`:
```
Apr 10 23:04:45 tp.flokli.de systemd[2252]: Started GnuPG cryptographic agent and passphrase cache.
Apr 10 23:04:45 tp.flokli.de gpg-agent[27465]: gpg-agent[27465]: reading options from '/home/flokli/.gnupg/gpg-agent.conf'
```

It also log if the file couldn't be found:
```
Apr 10 23:03:39 tp.flokli.de systemd[2252]: Started GnuPG cryptographic agent and passphrase cache.
Apr 10 23:03:39 tp.flokli.de gpg-agent[27209]: gpg-agent[27209]: Note: no default option file '/home/flokli/.gnupg/gpg-agent.conf'
```

That was with latest master - I'll check with 20.03 in a minute.

**Comment by flokli** [comment]
I can confirm the same behaviour with 20.03.

**Comment by flokli** [comment]
When I query via `gpgconf`, I also see the overidden value:
```
gpgconf --list-options gpg-agent  | grep 601
default-cache-ttl:24:0:expire cached PINs after N seconds:3:3:N:600::601
```

Can you explain what exactly doesn't work with your setup? I can't reproduce…

**Comment by matthew-piziak** [comment]
Hi @flokli, thank you so much for looking into this for me! I have just tested again now, and for whatever reason my system appears to be working now.

Previously, `gpgconf` was showing options that weren't in my `gpg-agent.conf`. But now, I can set my `gpg-agent.conf` and `gpgconf --list-options gpg-agent` immediately returns the correct values. Moreover I have double-checked with low TTL values and upon reloading the `systemctl` unit, I can see that those values are being respected.

Thank you especially for providing the exact commands you used to debug this; it's useful to have those at hand when I'm trying to follow along.

Everything looks good now. Closing. :)
